runner.py

```diff
@@ -0,0 +1,707 @@
+
+                # Make sure if parsed infos are as expected
+                fail_no = len([t for t in test_results if t.status == UnitTest.FAIL])
+                ignore_no = len([t for t in test_results if t.status == UnitTest.IGNORE])
+                if (fail_no != fail or ignore_no != ignore or len(test_results) != total):
+                    raise ValueError("Unit test harness: results do not match")
+
+                return test_results
+
+
+class TestCase:
+    PASS = "PASS"
+    FAIL = "FAIL"
+    FAIL_TIMEOUT = "FAIL TIMEOUT"
+    IGNORE = "IGNORE"
+
+    def __init__(self, name, target, syspage=None, status=None):
+        self.name = name
+        self.target = target
+        self.syspage = syspage
+        if not status:
+            status = TestCase.FAIL
+        self.status = status
+        self.harness = None
+
+    def __str__(self):
+        status = self.status
+        if status == TestCase.PASS:
+            color = Color.OK
+        elif status in (TestCase.FAIL, TestCase.FAIL_TIMEOUT):
+            color = Color.FAIL
+        elif status == TestCase.IGNORE:
+            color = Color.IGNORE
+
+        status = Color.colorify(self.status, color)
+        return f"{self.target}: {self.name}: {status}"
+
+    def ignored(self):
+        return self.status == TestCase.IGNORE
+
+    def failed(self):
+        return self.status == TestCase.FAIL
+
+    def passed(self):
+        return self.status == TestCase.PASS
+
+    def ignore(self):
+        self.status = TestCase.IGNORE
+
+    def fail(self):
+        self.status = TestCase.FAIL
+
+    def report(self):
+        logging.info(self)
+
+    def exec_syspage(self, proc):
+        if not self.syspage:
+            return
+
+        proc.expect_exact('(psh)% ')
+        proc.sendline(f'exec syspage/{self.syspage}')
+        proc.expect(f'exec syspage/{self.syspage}(.*)\n')
+
+    def handle(self, proc):
+        if self.ignored():
+            return True
+
+        self.status = TestCase.FAIL
+
+        if not self.harness:
+            return False
+
+        if self.syspage:
+            try:
+                self.exec_syspage(proc)
+            except pexpect.exceptions.TIMEOUT as exc:
+                logging.info(f"{self.target}: Executing syspage {self.syspage} failed!")
+                logging.debug(exc)
+                self.status = TestCase.FAIL_TIMEOUT
+                return False
+
+        res = False
+        try:
+            res = self.harness(proc)
+        except pexpect.exceptions.TIMEOUT as exc:
+            logging.debug(str(exc))
+            self.status = TestCase.FAIL_TIMEOUT
+        except pexpect.exceptions.EOF as exc:
+            logging.warning(f"Test {self.name} reached EOF")
+            logging.warning(str(exc))
+        except Exception as exc:
+            logging.warning(f"Test {self.name} exception:\n {exc}")
+            raise exc
+
+        if res:
+            self.status = TestCase.PASS
+
+        return res
```

Comment:
not sure, but return value is not being used (not being passed up from `DeviceRunner`?)

Reply:
For now I used `res` in `QemuRunner` to make decision if we should send `SIGKILL` or `SIGTERM` but now I think it won't be necessary.